Prepare GenAPI so it can be used for a new assembly diff generator

carlossanlop opened the pull request:
We intend to create a new tool to generate the monthly API diffs that will replace arcade's AsmDiff, which depends on Cci, an obsolete and unsupported library.

This PR makes the preliminary changes needed to be able to call GenAPI from yet another project besides GenAPI.Tool, GenAPI.Task and ApiCompat.

I made the following changes:

- Added a new class, GenApiAppConfiguration, that can be used to store the settings for GenAPI.
  - Make it an option to load the assemblies either from the filesystem or from streams representing the assembly code as string. Same for the assembly references.
  - Make it an option to set the header either from a file or from a string.
  - Move into this class the job of collecting the DocId files (either for API or attribute exclusion) from the filesystem. This way, the user can create a configuration instance passing a collection of DocID strings directly, not from disk.
  - Move into this class the code that configures the assemblies and the loader. It was repeated in GenApiApp and in the tests.
  - Use this class to allow making the output either the filesystem or streams for later usage.

The diffing tool will mainly need the last point: being able to have the strings stored in streams to be able to do the comparisons afterwards.

carlossanlop committed:
Prepare GenAPI so it can be used for a new assembly diff generator.

ViktorHofer commented on src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/Filtering/DocIdSymbolFilter.cs:
```diff
@@ -9,9 +9,15 @@ namespace Microsoft.DotNet.ApiSymbolExtensions.Filtering
-        private readonly HashSet<string> _docIdsToExclude = new(ReadDocIdsAttributes(docIdsToExcludeFiles));
+        private readonly HashSet<string> _docIdsToExclude;
+
+        public static DocIdSymbolFilter GetFilterForDocIds(params string[] docIdsToExclude)
```
Could this just be the class's constructor?

carlossanlop replied:
I guess it could, but I've seen that normally the preference is to have constructors be used for initialization only, and instance methods are used to do actual functionality. Plus that's what we were doing before, and the method name describes what it's doing (the constructor name does not).

ViktorHofer replied:
We prefer to use a static method when we do more work that what is expected in a constructor. Looking at the code here it looks like we are opening and reading files which definitely shouldn't happen in a constructor. So I agree, the static method makes more sense here.

IMO `Create` as the static method name makes more sense. Example: https://github.com/dotnet/sdk/blob/7361d854bbdb8681556ea579a5c1e2c4e1ffb051/src/Compatibility/ApiCompat/Microsoft.DotNet.PackageValidation/Package.cs#L123

In that case you invoke `Package.Create`. Here you would invoke `DoIdSymbolFilter.Create` which I think is also descriptive enough.

carlossanlop replied:
Alright, I can change it to Create.

carlossanlop replied:
Done.

ViktorHofer commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:
```diff
@@ -14,57 +13,41 @@
+        public CSharpFileBuilder(GenApiAppConfiguration configuration, TextWriter textWriter)
         {
-            _logger = logger;
+            _c = configuration;
```
Please use a more descriptive field name.

ViktorHofer commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:
```diff
@@ -14,57 +13,41 @@
+    public sealed class CSharpFileBuilder : IAssemblySymbolWriter
     {
-        private readonly ILog _logger;
+        private readonly GenApiAppConfiguration _c;
```
Please use a more descriptive field name.

carlossanlop replied:
Done.

ViktorHofer commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:
```diff
@@ -14,57 +13,41 @@
-            bool includeAssemblyAttributes,
-            IEnumerable<MetadataReference> metadataReferences)
+
+        public CSharpFileBuilder(GenApiAppConfiguration configuration, TextWriter textWriter)
```
Before we were passing-in explicitly what the class needed. Why do we need to change that?

carlossanlop replied:
Hmmm well, I see your point: on one hand, I don't intend to change the behavior of this class (I will create a new one), so the arguments don't need to be changed. I can still pass only what the class needs. On the other hand, passing the whole configuration object is handy.

I can revert this to only take the arguments it needs. I don't think it will affect the new diff code.

carlossanlop replied:
Yes. Done.

carlossanlop committed:
Pass to CSharpFileBuilder constructor only the necessary arguments instead of the whole config.
Rename the config instance variables to something more meaningful instead of c or _c.
Fix the build failure happening under #if NET.

carlossanlop committed:
Rename DocIdSymbolFilter.GetFilterForDocIds to DocIdSymbolFilter.Create.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/GenAPIConfiguration.cs:
```diff
@@ -0,0 +1,349 @@
+        private ILog? _logger = null;
+        private string[]? _assembliesPaths = null;
+        private string[]? _assemblyReferencesPaths = null;
+        private OutputType _outputType = OutputType.Console;
```
The output for the differ will be a single directory too.

It's only the input that will be two different directories.

I intend to generate two configuration instances, one for each input set. This configuration will come from the differ command line options, so I don't immediately see a problem.

But I see why you ask. Let me think about this more, maybe I can simplify it.

ericstj replied:
Well, there is a mix of things.  Some things need to be separate others shared.  This configuration class also contains a lot of things which aren't really configuration but have a lot more business logic.  I expect that you'll also end up with a different implementation of the business logic to drive the diff than the current GenAPI logic.

I wonder if you need a separate "Application" like type that's unique for AssmDiff that doesn't try to reuse the GenAPI "Application" and then focus on making the lower-level types better suit your needs?

carlossanlop replied:
Yes, I agree. I see the things that don't need to be shared. I advanced some more so I know what to reduce. Will push a new commit soon.

carlossanlop replied:
I addressed this by detaching from GenAPIConfiguration all the things that aren't shared between the differ and the existing GenAPIApp.

I also added static helper methods in that class for generating the things that are used in GenAPIApp, in the differ, and in the tests.

carlossanlop committed:
Load assemblies as dictionary instead of list, where the key is the assembly name.

ViktorHofer commented on src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/Filtering/DocIdSymbolFilter.cs:
```diff
@@ -9,9 +9,15 @@ namespace Microsoft.DotNet.ApiSymbolExtensions.Filtering
-        private readonly HashSet<string> _docIdsToExclude = new(ReadDocIdsAttributes(docIdsToExcludeFiles));
+        private readonly HashSet<string> _docIdsToExclude;
+
+        public static DocIdSymbolFilter Create(params string[] docIdsToExclude)
```
The parameter name here should indicate that this is about files. The previous parameter name was `docIdsToExcludeFiles`.

ViktorHofer commented on src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/Filtering/DocIdSymbolFilter.cs:
```diff
@@ -29,10 +35,15 @@ public bool Include(ISymbol symbol)
         }
 
-        private static IEnumerable<string> ReadDocIdsAttributes(IEnumerable<string> docIdsToExcludeFiles)
+        private static IEnumerable<string> ReadDocIdsAttributes(params string[] docIdsToExcludeFiles)
```
Consider merging this method into the static Create method above.

carlossanlop committed:
Simplify GenAPIConfiguration

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:
```diff
@@ -56,13 +61,21 @@ public CSharpFileBuilder(ILog logger,
 
         /// <inheritdoc />
         public void WriteAssembly(IAssemblySymbol assemblySymbol)
```
I'm thinking that the CSharpFileBuilder class should just have a WriteAssembly method. The functionality from the methods that are inherited, since it's shared between GenAPIApp, the differ and the tests, should probably be moved to a different class. To make it generic (C# or VB in the future) I might have to use an interface as well.

I'm going to think about this more.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/GenAPIConfiguration.cs:
```diff
@@ -0,0 +1,317 @@
+        }
+
+
+        private static IEnumerable<KeyValuePair<string, ReportDiagnostic>> DiagnosticOptions { get; } = new[]
```
All the code under this line is duplicate from a test extension class. I need to consolidate it.

ViktorHofer commented on test/Microsoft.DotNet.ApiCompatibility.Tests/Microsoft.DotNet.ApiCompatibility.Tests.csproj:
```diff
@@ -16,8 +16,9 @@
 
   <ItemGroup>
-    <ProjectReference Include="$(RepoRoot)src\Compatibility\ApiCompat\Microsoft.DotNet.ApiCompatibility\Microsoft.DotNet.ApiCompatibility.csproj" />
+    <ProjectReference Include="$(RepoRoot)src\Compatibility\GenAPI\Microsoft.DotNet.GenAPI\Microsoft.DotNet.GenAPI.csproj" />
```
revert:

```suggestion
```

carlossanlop replied:
You found it! This is why I couldn't figure out where all the ApiCompat failures were coming from in the tests. It got lost in the weeds.

carlossanlop committed:
Use factory methods instead of fluent. Get rid of GenAPIConfiguration, split it into two separate classes.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI.Task/GenAPITask.cs:
```diff
@@ -62,17 +64,21 @@ public class GenAPITask : TaskBase
         /// <inheritdoc />
         protected override void ExecuteCore()
         {
+            (IAssemblySymbolLoader loader, Dictionary<string, IAssemblySymbol> assemblySymbols) = AssemblyLoaderFactory.CreateFromFiles(
+                assembliesPaths: Assemblies ?? throw new NullReferenceException("Assemblies cannot be null."),
+                assemblyReferencesPaths: AssemblyReferences,
+                RespectInternals);
```
I replaced the GenAPIConfiguration with AssemblyLoaderFactory. Only static methods.

The GenAPIConfiguration class also used to contains the methods for symbol filter generation. I created a separate factory for that.

ViktorHofer replied:
One of the main design principles in the Compatibility tools is to keep the UI layer as thin as possible and let the business logic layer handle with the start-up and shutdown. The existing code base does this well, it just calls into GenAPI.Run without calling into lower level types like inside `Microsoft.CodeAnalysis` or `ApiSymbolExtensions`.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI.Task/GenAPITask.cs:
```diff
@@ -62,17 +64,21 @@ public class GenAPITask : TaskBase
+                          ExcludeApiFiles,
+                          ExcludeAttributesFiles,
+                          RespectInternals,
+                          IncludeAssemblyAttributes);
```
The arguments remained mostly the same in GenAPIApp.Run. The only two changes were the first two parameters: I need to pass the loader and the IAssemblySymbol dictionary now. Those will need to be processed beforehand because it's the code I need to share with ApiDiff.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpFileBuilder.cs:
```diff
@@ -1,348 +1,72 @@
     /// Processes assembly symbols to build corresponding structures in C# language.
     /// </summary>
-    public sealed class CSharpFileBuilder : IAssemblySymbolWriter, IDisposable
+    public sealed class CSharpFileBuilder : IAssemblySymbolWriter
```
I extracted all the CSharpFileBuilder logic that visited nodes and put it in a separate class calls CSharpAssemblyVisitor (and its corresponding IAssemblyVisitor, in case we ever do the same for VB). I did this because that's the code I need to share with ApiDiff.

The only code left in CSharpFileBuilder is the code that writes the assembly to disk (after calling the necessary CSharpAssemblyVisitor methods).

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/GenAPIApp.cs:
```diff
@@ -32,58 +30,20 @@ public static void Run(ILog logger,
             bool respectInternals,
             bool includeAssemblyAttributes)
         {
-            bool resolveAssemblyReferences = assemblyReferences?.Length > 0;
```
This logic that creates the loader is what I need to share with ApiDiff, so it's been moved up to AssemblyLoaderFactory. So the loader and the dictionary of IAssemblySymbols are now being passed as the first two arguments.

ericstj commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/AssemblyLoaderFactory.cs:
```diff
@@ -0,0 +1,36 @@
+
+namespace Microsoft.DotNet.GenAPI;
+
+public class AssemblyLoaderFactory
```
Can you add comments to public types?

ericstj reviewed (commented):
A couple small comments, would like for @ViktorHofer to weigh in on style / architecture before going deeper.

carlossanlop committed:
Address feedback and make some other improvements.

ViktorHofer commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/SyntaxRewriter/TypeDeclarationCSharpSyntaxRewriter.cs:
```diff
@@ -17,6 +17,14 @@ namespace Microsoft.DotNet.GenAPI.SyntaxRewriter
+        /// Initializes a new instance of the <see cref="TypeDeclarationCSharpSyntaxRewriter"/> class, and optionally allows deciding whether to insert the partial modifier for types or not.
+        /// </summary>
+        /// <param name="addPartialModifier">Determines whether to insert the partial modifier for types or not.</param>
+        public TypeDeclarationCSharpSyntaxRewriter(bool addPartialModifier = true) => _addPartialModifier = addPartialModifier;
```
It would be better to not set a default here and instead let the business logic explicitly provide the right setting. The partial modifier is only useful when generating code that should be compile-able and extend-able which I see very specific to GenAPI.

carlossanlop replied:
Alright.

ViktorHofer commented on src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs:
```diff
@@ -256,6 +280,40 @@ public IEnumerable<IAssemblySymbol> LoadMatchingAssemblies(IEnumerable<IAssembly
         }
 
+        /// <inheritdoc />
+        public void LogAllDiagnostics(ILog logger, string? customMessage = null)
```
I think `headerMessage` would be a better arg name

ViktorHofer commented on src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/AssemblySymbolLoader.cs:
```diff
@@ -256,6 +280,40 @@ public IEnumerable<IAssemblySymbol> LoadMatchingAssemblies(IEnumerable<IAssembly
         }
 
+        /// <inheritdoc />
+        public void LogAllDiagnostics(ILog logger, string? customMessage = null)
```
The logger should be an instance variable. This type would benefit from using the logger more extensively. I was surprised to see that isn't yet used by this type. It should be passed through the assembly symbol factory's constructor.

carlossanlop replied:
I addressed this, but just FYI, it affected the pre-existing IAssemblySymbolLoaderFactory and AssemblySymbolLoaderFactory types. I'll push my updates in a few moments.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/AssemblyLoaderFactory.cs:
```diff
@@ -0,0 +1,36 @@
+
+public class AssemblyLoaderFactory
+{
+    public static (IAssemblySymbolLoader, Dictionary<string, IAssemblySymbol>) CreateFromFiles(string[] assembliesPaths, string[]? assemblyReferencesPaths, bool respectInternals = false)
```
Sure, no problem.

BTW I just realized there is a pre-existing class called AssemblySymbolLoaderFactory. It is not a static class as it inherits from an interface called IAssemblySymbolLoaderFactory which has only one method to create a default AssemblySymbolLoader instance. Mentioning as my new class (which I plan to remove) had a similar intention (inherit from its own interface too so it was independent from the language), but with the purpose of creating a more advanced/complete instance of IAssemblySymbolLoader.

carlossanlop replied:
I'll have to modify that class a bit since I was asked in another comment to take care of passing the ILog instance properly to the AssemblySymbolLoader (we weren't doing that before my PR).

carlossanlop committed:
Remove unnecessary factory class and pass logger where it was missing.

carlossanlop committed:
Merge CompositeSymbolFilter changes with main

carlossanlop committed:
Do not set addPartialModifier to true by default in TypeDeclarationCSharpSyntaxRewriter

ericstj commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpAssemblyVisitor.cs:
```diff
@@ -0,0 +1,362 @@
+    }
+
+    /// <inheritdoc />
+    public SyntaxNode GetFormattedRootNodeForDocument(Document document) => document.GetSyntaxRootAsync().Result!.Rewrite(new SingleLineStatementCSharpSyntaxRewriter());
```
Does this actually use any instance state, or is it just a transformation on Document?  Why does this live on the assembly visitor?

carlossanlop replied:
I renamed the class as suggested so the purpose of this method made more sense.

ericstj replied:
Just tossing the idea out there - now that you have both CSharpAssemblyDocumentGenerator and CSharpFileBuilder do you think it makes sense to have these two different types?  Could we combine them back together and just have one that is used slightly differently depending on the caller?  It might reduce the total number of names we have to come up with while still being meaningful and keeping the diff smaller.

ericstj commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/IAssemblyVisitor.cs:
```diff
@@ -0,0 +1,23 @@
+
+namespace Microsoft.DotNet.GenAPI;
+
+public interface IAssemblyVisitor
```
Can you document this, and mention its purpose?  Is it for testability - and if so, do you use that anywhere?  For that matter I also see that `IAssemblySymbolWriter` is an interface only ever used with a single type.  @ViktorHofer  do you think there is value in keeping these interfaces?

carlossanlop replied:
Correct me if I'm wrong, but we use interfaces in case we one day decide to provide similar support to Visual Basic code, no?

ericstj replied:
I'm not certain, but if we did add VB support we could add the interface at that time?  I tried to look back at the history and it was there from the start.  I couldn't see anywhere we actually made use of it.

ViktorHofer replied:
> I also see that IAssemblySymbolWriter is an interface only ever used with a single type. @ViktorHofer do you think there is value in keeping these interfaces?

Yes, the general idea for `IAssemblySymbolWriter` was to not depend on the specific implementation in the unit of code that depends on it (basically follow the Liskov Substitution Principle). Looks like GenAPIApp depends on the concrete type though so the interface doesn't serve much purpose today. There's probably room for improvement here and I would keep the interface.

ericstj commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/CSharpAssemblyVisitor.cs:
```diff
@@ -0,0 +1,341 @@
+
+namespace Microsoft.DotNet.GenAPI;
+
+public class CSharpAssemblyVisitor : IAssemblyVisitor
```
nit: sealed.
nit: naming.  This doesn't actually expose a Visitor pattern so naming it a Visitor seems off.  It generates a C# document of API signatures for an assembly.  Maybe CSharpAssemblySignatureGenerator or CSharpAssemblyApiGenerator?

carlossanlop replied:
CSharpAssemblyDocumentGenerator sounds like a better name.

I need to think what to do about the IAssemblyVisitor interface this inherits from.

carlossanlop replied:
Hold on I am wrong - The second class I mentioned is not implementing this interface. There's only one class implementing it, the one I just renamed. I think we can remove the interface altogether.

carlossanlop commented on src/Compatibility/GenAPI/Microsoft.DotNet.GenAPI/SymbolFilterFactory.cs:
```diff
@@ -0,0 +1,86 @@
+
+namespace Microsoft.DotNet.GenAPI;
+
+public static class SymbolFilterFactory
```
I think it's possible. Let me explore it and see if it makes sense.

carlossanlop committed:
Address latest batch of feedback: Rename the wrongly named CSharpAssemblyVisitor class, remove its interface. Move the static methods from SymbolFilterFactory to CompositeSymbolFilter, and delete SymbolFilterFactory. Move TryGetRecordConstructor static extension method to a higher assembly. Move the NullableAttributes.cs dependency to the same higher assembly. Move ImplicitSymbolFilter to a higher assembly. Adjust the tests accordingly.

ViktorHofer commented on src/Compatibility/Microsoft.DotNet.ApiSymbolExtensions/Filtering/CompositeSymbolFilter.cs:
```diff
@@ -14,7 +14,114 @@ public sealed class CompositeSymbolFilter(params IEnumerable<ISymbolFilter> filt
+                                                                     bool includeEffectivelyPrivateSymbols = true,
+                                                                     bool includeExplicitInterfaceImplementationSymbols = true)
+        {
+            DocIdSymbolFilter? docIdSymbolFilter =
```
The CompositeSymbolFilter is _one_ implementation of the ISymbolFilter interface and must not know about others, i.e. the `DocIdSymbolFilter`. Even if this these methods are static, they don't belong here. That would violate the separation of concerns.

carlossanlop replied:
I'll bring back the factory class. If there are methods that belong to a specific class, I'll put them there instead.

ViktorHofer reviewed (commented):
This PR is getting quite big and complex. I think it would help if you would submit some of the non-controversial changes as one separate PR. Examples to include:
- AssemblySymbolLoader receiving the log instance
- The ctor change in DocIdSymbolFilter
- and the whitespace removals and typo fixes

carlossanlop commented:
I'll submit smaller PRs as this one is already full of merge conflicts.
